src/interface.jl

```diff
@@ -201,16 +201,19 @@ isring(geom) = isclosed(geom) && issimple(geom)
 Return the length of `geom` in its 2d coordinate system.
 Note that this is only valid for [`AbstractCurveTrait`](@ref)s.
 """
-length(geom) = length(geomtrait(geom), geom)
+length(geom) = length(crstrait(geom), geomtrait(geom), geom)
+length(::UnknownTrait, trait, geom) = length(trait, geom)  # fallback
 
 # Surface
+
 """
     area(geom) -> Number
 
 Return the area of `geom` in its 2d coordinate system.
 Note that this is only valid for [`AbstractSurfaceTrait`](@ref)s.
 """
-area(geom) = area(geomtrait(geom), geom)
+area(geom) = area(crstrait(geom), geomtrait(geom), geom)
+area(::UnknownTrait, trait, geom) = area(trait, geom)  # fallback
```

Comment:
Should this be AbstractProjectedTrait too?

Reply:
I had that first, but as soon as a package implements area with ProjectedTrait, this fallback is not needed.

Reply:
I guess this goes back to "which is the blessed implementation", though - do we want packages dispatching on their own geometry types?

Reply:
I don't think blessed has anything to do with this. To implement GeoInterface, packages already have to dispatch on their own geometry type, it's part of the spec?

Reply:
But most geometry types are not going to implement these `area` algs right? Like `GeometryOps.area` already works for anything. 

I don't really get the use-case for these traits and `GeoInterface.area`. What package can and would implement the methods?

Reply:
I can see GEOS warning on geodesic trait at least, and some future `S2.jl` erroring on projected trait, but that seems like something that should be done in the package, not as a GeoInterface dispatch...

Reply:
> I don't really get the use-case for these traits and `GeoInterface.area`. What package can and would implement the methods?

I don't really get the question. We did discuss this extensively back at SDSL no? And there was agreement in #119 that this is necessary? I have implemented your own suggestion of an UnknownTrait() in this PR.

Essentially, we mix all kind of geometries in our ecosystem, sometimes projected, sometimes geographic. Our backing implementations (either via GEOS or GeometryOps) are correct for projected only, and will give plain wrong answers on anything geographic. So there's a need for the distinction, and a trait here seems like the logical choice. This PR is fully backwards compatible, but depwarns (so for devs only) that a distinction between project and geographic should be implemented for certain GeoInterface methods that are sensitive to it.

> But most geometry types are not going to implement these area algs right? Like GeometryOps.area already works for anything.

Except for geographic things, and that's ignoring anything already in ArchGDAL, LibGEOS and derivatives. And things like extent, which even has a fallback here (again, only valid for projected data).

> I can see GEOS warning on geodesic trait at least, and some future `S2.jl` erroring on projected trait, but that seems like something that should be done in the package, not as a GeoInterface dispatch...

This package is an interface definition, so yes by definition the package has to implement it (and in doing so determine whether it errors/warns/ignores). Any package that implements GeoInterface just has to implement crstrait(geom) = ProjectedTrait() to make the depwarn go away (which a dev only see on tests, nothing in this PR is seen by an end user).

Reply:
But that's all orthogonal to `GeoInterface.area` needing a method for it. I'm not sure what package would add that method or why

Reply:
Are we only talking about `area`, or also about `length`, `extent`, etc? The `area` was actually the example discussed in #119, here I assume we also mean the others.

If we accept we need a crstrait because the required algorithm to implement `area` is different for projected vs geographic geoms, I'd say it follows you want to dispatch on it (it's a trait after all). As this is the *interface* package, it also prescribes the method needed to implement it. Just like we prescribe that all methods do `thing(x) = thing(trait(x), x)`, and packages implement the latter, not the former, even when the package itself does not always need it.

> I'm not sure what package would add that method or why

All packages that implement GeoInterface should eventually switch from `area(::geomtrait, geom)` to `area(::crstrait, ::geomtrait, geom)`, to make explicit what kind of crstrait(s) they support for these methods. With the current PR the default is Projected, so geometries that are Geographic in nature will not work with the current `area` implementations. 

It is actually a very small (and imho elegant) change, especially if you consider the alternative that each package starts adding its own dispatch method/if/else mechanism to check the crstrait, or worse they do not do so and are silently wrong as they currently are.

I believe we require this distinction before we can even think about implementing GeoInterface for https://github.com/TLipede/S2Geometry.jl, https://github.com/wookay/H3.jl, https://github.com/anowacki/GeographicLib.jl, or the geodesic part of Proj.jl.

Reply:
My point is most packages that own geometries don't implement `GeoInterface.area` and probably won't ever implement `area`.

Personally I mostly have GeoInterface.jl / GeoJSON.jl / Shapefile.jl or in future GeoPackage.jl objects. How and where can they dispatch to implemented `GeoInterface.area` methods?

I just can't see the practical end goal.

Reply:
But that's a critique on having area in GeoInterface at all right? And more related to blessed implementations than crstraits?

Reply:
Yes, but I would personally deprecate it rather than modifying it in this PR.

(I do actually want to hear the vision for how it will work for those types... Maybe at JuliaCon)

Reply:
I can understand that, but as we don't have a blessed design yet, it's hard to deprecate anything. Perfect is the enemy of good, and I think this PR is an improvement over the current state.

And in terms of usage, I think the extent (and specifically the calc_extent fallback here), and everything that uses ArchGDAL (as it has both widely used geometries and has an area implementation) would be improved by this PR.

Reply:
Ok, so I've reverted all of this for the operations, except for extent (since we have a fallback to calc_extent in there). Ideally we would move that to GeometryOps, but for now we should leave it in (would also be breaking I guess?)